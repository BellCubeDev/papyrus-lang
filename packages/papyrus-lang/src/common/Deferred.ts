export class Deferred<T> {
    private readonly _promise: Promise<T>;

    private _resolve: (value?: T | PromiseLike<T>) => void;
    private _reject: (reason?: any) => void;

    constructor() {
        // tslint:disable-next-line:promise-must-complete
        this._promise = new Promise<T>((resolve, reject) => {
            this._resolve = resolve;
            this._reject = reject;
        });
    }

    public resolve(value?: T | PromiseLike<T>) {
        this._resolve(value);
    }

    public reject(reason?: any) {
        this._reject(reason);
    }

    get promise() {
        return this._promise;
    }
}
