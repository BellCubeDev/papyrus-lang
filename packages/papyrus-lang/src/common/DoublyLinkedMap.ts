export class DoublyLinkedMap<TKey, TValue> {
    get keys() {
        return this._keyValue.keys();
    }

    get values() {
        return this._valueKey.keys();
    }
    public static fromMap<TKey, TValue>(
        map: ReadonlyMap<TKey, TValue>,
        transformKey: (key: TKey) => TKey
    ) {
        const newMap = new DoublyLinkedMap<TKey, TValue>();

        for (const entry of map.entries()) {
            const transformedKey = transformKey
                ? transformKey(entry[0])
                : entry[0];
            newMap.set(transformedKey, entry[1]);
        }

        return newMap;
    }

    private readonly _keyValue: Map<TKey, TValue> = new Map();
    private readonly _valueKey: Map<TValue, TKey> = new Map();

    public getValueFromKey(key: TKey): TValue {
        return this._keyValue.get(key);
    }

    public getKeyFromValue(value: TValue): TKey {
        return this._valueKey.get(value);
    }

    public hasValue(value: TValue) {
        return this._valueKey.has(value);
    }

    public hasKey(key: TKey) {
        return this._keyValue.has(key);
    }

    public set(key: TKey, value: TValue) {
        this._keyValue.set(key, value);
        this._valueKey.set(value, key);
    }

    public deleteByKey(key: TKey) {
        if (this._keyValue.has(key)) {
            const value = this._keyValue.get(key);
            this._valueKey.delete(value);
            this._keyValue.delete(key);
        }
    }

    public deleteByValue(value: TValue) {
        if (this._valueKey.has(value)) {
            const key = this._valueKey.get(value);
            this._keyValue.delete(key);
            this._valueKey.delete(value);
        }
    }

    public clear() {
        this._keyValue.clear();
        this._valueKey.clear();
    }
}
