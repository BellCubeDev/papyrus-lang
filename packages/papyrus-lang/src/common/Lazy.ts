export class Lazy<T> {
    private readonly _factory: () => T;
    private readonly _expired: (value: T) => boolean;
    private _value: T;

    constructor(
        factory: () => T,
        expired: (value: T) => boolean = () => false
    ) {
        this._factory = factory;
        this._expired = expired;
    }

    get value() {
        if (typeof this._value === 'undefined' || this._expired(this._value)) {
            this._value = this._factory();
        }

        return this._value;
    }

    get currentValue() {
        return this._value;
    }

    public refresh() {
        this._value = this._factory();
    }
}

export class LazyTokenEqualityMemoized<T> extends Lazy<T> {
    private _token: any;

    constructor(factory: () => T, getToken: () => any) {
        super(factory, () => {
            const newToken = getToken();
            if (newToken !== this._token) {
                this._token = newToken;
                return true;
            }

            return false;
        });
    }
}
