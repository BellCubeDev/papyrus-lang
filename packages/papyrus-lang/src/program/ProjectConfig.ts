import { xml2js } from 'xml-js';

export enum AssemblyMode {
    none = 'none',
    keep = 'keep',
    only = 'only',
    discard = 'discard',
}

export interface ProjectConfig {
    filePath?: string;
    output?: string;
    flags?: string;
    asm?: AssemblyMode;
    optimize?: boolean;
    release?: boolean;
    final?: boolean;
    imports: string[];
    folder: { noRecurse?: boolean; path: string };
    scripts?: string[];
}

function getEmptyProject(): ProjectConfig {
    return {
        output: null,
        flags: null,
        asm: 'none' as AssemblyMode,
        optimize: false,
        release: false,
        final: false,
        imports: null,
        folder: { noRecurse: false, path: '.' },
        scripts: null,
    };
}

interface XmlProjectAttributes {
    Flags: string;
    Output: string;
    Asm: AssemblyMode;
    Optimize: string;
    Release: string;
    Final: string;
}

interface XmlFolderAttributes {
    NoRecurse: string;
}

interface XmlImport {
    _text: string[];
}

interface XmlImports {
    Import: XmlImport[];
}

interface XmlFolder {
    _attributes: XmlFolderAttributes;
    _text: string[];
}

interface XmlFolders {
    Folder: XmlFolder[];
}

interface XmlScript {
    _text: string[];
}

interface XmlScripts {
    Script: XmlScript[];
}

interface XmlPapyrusProject {
    _attributes: XmlProjectAttributes;
    Imports: XmlImports[];
    Folders: XmlFolders[];
    Scripts: XmlScripts[];
}

interface XmlRootObject {
    PapyrusProject: XmlPapyrusProject[];
}

function parseBool(bool: string) {
    return JSON.parse(bool.toString().toLowerCase());
}

export function parsePapyrusProjectXml(xml: string): ProjectConfig {
    const parsed = xml2js(xml, {
        alwaysArray: true,
        compact: true,
        ignoreAttributes: false,
    }) as XmlRootObject;

    const project: ProjectConfig = { ...getEmptyProject() };

    const xmlProject = parsed.PapyrusProject[0];
    const projectAttributes = xmlProject._attributes;
    const imports = xmlProject.Imports ? xmlProject.Imports[0] : null;

    project.imports =
        imports && imports.Import ? imports.Import.map((i) => i._text[0]) : [];

    const folders = xmlProject.Folders ? xmlProject.Folders[0] : null;
    project.folder =
        folders && folders.Folder && folders.Folder.length > 0
            ? {
                  noRecurse:
                      folders.Folder[0]._attributes &&
                      folders.Folder[0]._attributes.NoRecurse
                          ? parseBool(folders.Folder[0]._attributes.NoRecurse)
                          : false,
                  path: folders.Folder[0]._text[0],
              }
            : null;

    const scripts = xmlProject.Scripts ? xmlProject.Scripts[0] : null;
    project.scripts =
        scripts && scripts.Script ? scripts.Script.map((s) => s._text[0]) : [];

    if (projectAttributes) {
        if (projectAttributes.Flags) {
            project.flags = projectAttributes.Flags;
        }

        if (projectAttributes.Asm) {
            project.asm = projectAttributes.Asm;
        }

        if (projectAttributes.Final) {
            project.final = parseBool(projectAttributes.Final);
        }

        if (projectAttributes.Optimize) {
            project.optimize = parseBool(projectAttributes.Optimize);
        }

        if (projectAttributes.Output) {
            project.output = projectAttributes.Output;
        }

        if (projectAttributes.Release) {
            project.release = parseBool(projectAttributes.Release);
        }
    }

    return project;
}
