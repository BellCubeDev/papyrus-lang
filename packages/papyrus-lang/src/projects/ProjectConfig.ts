export enum AssemblyMode {
    none = 'none',
    keep = 'keep',
    only = 'only',
    discard = 'discard',
}

export interface ProjectConfig {
    filePath?: string;
    output?: string;
    flags?: string;
    asm?: AssemblyMode;
    optimize?: boolean;
    release?: boolean;
    final?: boolean;
    imports: string[];
    folder: { noRecurse?: boolean; path: string };
    scripts?: string[];
}

export function createEmptyConfig(): ProjectConfig {
    return {
        output: null,
        flags: null,
        asm: 'none' as AssemblyMode,
        optimize: false,
        release: false,
        final: false,
        imports: null,
        folder: { noRecurse: false, path: '.' },
        scripts: null,
    };
}
