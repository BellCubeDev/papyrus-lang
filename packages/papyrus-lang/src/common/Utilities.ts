import { Deferred } from './Deferred';
import { Range } from './Range';
import { IterableScannableIterator, Scanner } from './Scanner';

export function delayAsync(durationMs: number = 0) {
    const deferred = new Deferred<void>();
    setTimeout(() => deferred.resolve(), durationMs);
    return deferred.promise;
}

export function createIterator<T>(iterable: Iterable<T>) {
    return iterable[Symbol.iterator]();
}

export function* iterateWhere<T>(
    iterable: Iterable<T>,
    filter: (element: T) => boolean
) {
    for (const element of iterable) {
        if (filter(element)) {
            yield element;
        }
    }
}

export function* iterateMany<T>(
    iterableOfIterables:
        | Iterable<Iterable<T>>
        | Iterable<T[]>
        | Iterable<T>[]
        | T[][]
): IterableIterator<T> {
    for (const iterable of iterableOfIterables) {
        for (const element of iterable) {
            yield element;
        }
    }
}

export function lastOfIterable<T>(iterable: Iterable<T>) {
    let value: T = null;

    for (const current of iterable) {
        value = current;
    }

    return value;
}

export function* mapIterable<T, U>(
    iterable: Iterable<T>,
    map: (value: T) => U
) {
    for (const element of iterable) {
        yield map(element);
    }
}

export function clamp(value: number, min: number, max: number) {
    return Math.min(Math.max(value, min), max);
}

export function asScannableIterator<T>(
    iterator: Iterator<T>
): IterableScannableIterator<T> {
    if (iterator instanceof Scanner) {
        return iterator;
    }

    return new Scanner<T>(iterator);
}

export function* iterateNumbersForStartAndCount(start: number, count: number) {
    let value = start;
    const end = start + count;

    while (value < end) {
        yield value;
        value++;
    }
}

export function iterateNumbersForRange(range: Range) {
    return iterateNumbersForStartAndCount(range.start, range.start + range.end);
}

export function* iteratorOf<T>(value: T, count: number) {
    for (const _num of iterateNumbersForStartAndCount(0, count)) {
        yield value;
    }
}

export function stringOf(value: string, count: number) {
    return Array.from(iteratorOf(value, count)).join('');
}

export function flushIterator(iterator: Iterator<any>) {
    while (!iterator.next().done) {}
}
