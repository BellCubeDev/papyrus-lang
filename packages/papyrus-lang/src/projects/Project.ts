import * as path from 'upath';
import URI from 'vscode-uri';

import { IFileSystem } from '../host/FileSystem';
import { ProjectConfig } from './ProjectConfig';

export class Project {
    private readonly _fileSystem: IFileSystem;
    private readonly _config: ProjectConfig;

    constructor(config: ProjectConfig, @IFileSystem fileSystem: IFileSystem) {
        this._config = config;
        this._fileSystem = fileSystem;
    }

    get config() {
        return this._config;
    }

    get filePath() {
        return this._config.filePath;
    }

    public resolveFiles(): ReadonlyMap<string, string> {
        const groups = [
            ...[...this._config.imports]
                .reverse()
                .map((i) => this.findPscFiles(i, true)),
                this._config.folder ? this.findPscFiles(
                this._config.folder.path,
                !this._config.folder.noRecurse
            ) : [] as any,
        ];

        const files = new Map<string, string>();

        for (const group of groups) {
            for (const file of group.files) {
                files.set(file[1], file[0]);
            }
        }

        return files;
    }

    private findPscFiles(folderUri: string, recurse: boolean) {
        const folderPath = URI.parse(folderUri).fsPath;

        const paths = this._fileSystem.findFilesAsUris(
            path.join(folderPath, recurse ? '**' : '.', '*.psc')
        );

        return {
            basePath: folderPath,
            files: paths.map((f) => [
                f,
                path
                    .basename(
                        path.relative(folderPath, URI.parse(f).fsPath).replace(/\/|\\/g, ':'),
                        '.psc'
                    )
            ]),
        };
    }
}
