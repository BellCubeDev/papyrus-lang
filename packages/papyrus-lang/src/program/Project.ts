import * as path from 'upath';
import URI from 'vscode-uri';

import { findFiles, iterateMany, readTextFile } from '../common/Utilities';
import { parsePapyrusProjectXml, ProjectConfig } from './ProjectConfig';

export function findProjectFiles(dirPath: string) {
    return findFiles(path.normalize(path.join(dirPath, '**', '*.ppj')));
}

export function findProjectFilesInDirectories(dirPaths: string[]) {
    return Array.from(
        new Set(iterateMany<string>(dirPaths.map(findProjectFiles))).values()
    );
}

export function loadProjectFile(filePath: string) {
    const resolvedPath = path.resolve(filePath);
    const base = path.dirname(resolvedPath);
    const projectXml = readTextFile(filePath);

    const projectConfig = parsePapyrusProjectXml(projectXml);
    projectConfig.folder.path = path.resolve(
        base,
        path.normalizeSafe(projectConfig.folder.path)
    );

    projectConfig.imports = projectConfig.imports
        .map((importPath) => path.normalizeSafe(importPath))
        .map(
            (importPath) =>
                path.isAbsolute(importPath)
                    ? importPath
                    : path.resolve(base, importPath)
        );

    projectConfig.filePath = filePath;

    if (projectConfig.output) {
        projectConfig.output = path.normalizeSafe(projectConfig.output);
    }

    return projectConfig;
}

export class Project {
    private readonly _config: ProjectConfig;

    constructor(config: ProjectConfig) {
        this._config = config;
    }

    get filePath() {
        return this._config.filePath;
    }

    public resolveFiles(): ReadonlyMap<string, string> {
        const groups = [
            ...[...this._config.imports]
                .reverse()
                .map((i) => this.findPscFiles(path.normalizeSafe(i), true)),
            this.findPscFiles(
                path.normalizeSafe(this._config.folder.path),
                !this._config.folder.noRecurse
            ),
        ];

        const files = new Map<string, string>();

        for (const group of groups) {
            for (const file of group.files) {
                files.set(file[1], URI.file(file[0]).toString());
            }
        }

        return files;
    }

    private findPscFiles(folderPath: string, recurse: boolean) {
        const paths = findFiles(
            path.join(folderPath, recurse ? '**' : '.', '*.psc')
        );

        return {
            basePath: folderPath,
            files: paths.map((f) => [
                f,
                path.basename(
                    path.relative(folderPath, f).replace(/\/|\\/g, ':'),
                    '.psc'
                ),
            ]),
        };
    }
}
