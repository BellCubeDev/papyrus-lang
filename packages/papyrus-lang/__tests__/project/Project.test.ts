import * as path from 'path';
import { readTextFile } from '../../src/common/Utilities';
import { loadProjectFile, Project } from '../../src/program/Project';
import { parsePapyrusProjectXml } from '../../src/program/ProjectConfig';

describe('Project', () => {
    describe('parsePapyrusProjectXml', () => {
        it('parses a ppj file', () => {
            const source = readTextFile(
                path.resolve(
                    __dirname,
                    '../../../../papyrus/FO4TestScripts/Project/Project.ppj'
                )
            );
            const project = parsePapyrusProjectXml(source);
            expect(project).toMatchSnapshot();
        });
    });

    describe('loadProjectFile', () => {
        it('parses a ppj file and creates a config with resolved paths', () => {
            loadProjectFile(
                path.resolve(
                    __dirname,
                    '../../../../papyrus/FO4TestScripts/Project/Project.ppj'
                )
            );
        });
    });

    // it('resolves files with the correct precedence', () => {
    //     const projectConfig = loadProjectFile(
    //         path.resolve(__dirname, '../../../../papyrus/FO4TestScripts/Project/Project.ppj')
    //     );
    //     const project = new Project(projectConfig);
    //     const files = Array.from(project.resolveFiles().entries());

    //     expect(files[0][1]).not.toBe(files[1][1]);
    // });
});
