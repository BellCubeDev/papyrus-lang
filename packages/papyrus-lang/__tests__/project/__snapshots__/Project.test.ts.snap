// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project parsePapyrusProjectXml parses a ppj file 1`] = `
Object {
  "asm": "none",
  "final": false,
  "flags": "Institute_Papyrus_Flags.flg",
  "folder": Object {
    "noRecurse": false,
    "path": ".",
  },
  "imports": Array [
    "C:\\\\Program Files (x86)\\\\Steam\\\\steamapps\\\\common\\\\Fallout 4\\\\Data\\\\Scripts\\\\Source\\\\Base",
    "..\\\\..\\\\FO4Scripts\\\\Base",
  ],
  "optimize": false,
  "output": null,
  "release": false,
  "scripts": Array [],
}
`;
